[SimpleLoopUnswitch] Record loops from unswitching non-trivial conditions

antoniofrighetto opened the pull request:
Track newly-cloned loops coming from unswitching non-trivial invariant conditions, so as to prevent conditions in such cloned blocks from being unswitched again.

Fixes: https://github.com/llvm/llvm-project/issues/138509.

dtcxzyw commented on llvm/lib/Transforms/Scalar/SimpleLoopUnswitch.cpp:
```diff
@@ -2142,34 +2142,33 @@ void visitDomSubTree(DominatorTree &DT, BasicBlock *BB, CallableT Callable) {
-  if (!NewLoops.empty())
+  auto RecordLoopAsUnswitched = [&](Loop *TargetLoop, StringRef Tag) {
+    auto &Ctx = TargetLoop->getHeader()->getContext();
+    const auto &DisableMDName = (Twine(Tag) + ".disable").str();
```
`Twine::str` returns a temporary `std::string`. Can we add a parameter `StringRef DisableTag`  instead?

antoniofrighetto replied:
Yeah, the temporary `std::string`, being bound to a `const std::string&` (which extends its lifetime), was a tradeoff in favour of reducing verbosity. Switched to passing `StringRef` as parameter.

dtcxzyw reviewed (approved):
LGTM. Thank you!
Can you test this patch on llvm-test-suite to ensure that it doesn't cause performance regressions?

antoniofrighetto committed:
[SimpleLoopUnswitch] Record loops from unswitching non-trivial conditions

Track newly-cloned loops coming from unswitching non-trivial invariant
conditions, so as to prevent conditions in such cloned blocks from
being unswitched again.

Fixes: https://github.com/llvm/llvm-project/issues/138509.

nikic commented:
Does this mean that we can now only unswitch a single condition in the loop?

bjope commented:
For the record, this did cause benchmark regressions for our downstream target. Worst case I've seen was a cycle increase by a factor 1.3 for a benchmark. I don't have much more to say about it right now (haven't had time to look at it closer to understand the exact scenario). Just wanted to mention this, if it turns out more people have problems.

bjope commented:
Removing it at the end of the "pass" is perhaps not as simple as it sounds as it rather would be at the end of the loop pass manager or something like that (given that the loop pass manager is running the pass on one loop at a time).
